### PR TITLE
Fix :nth-child selector

### PR DESCRIPTION
**Pull request by mauke**
### Summary
Handle `:nth-child(An+B)` algebraically instead of iterating over possible values of n.

### Motivation
Previous implementation was broken (but so were the tests, so nobody noticed).

### References
See issue #2023.

### CODE REVIEW

File: t/mojo/dom.t [hunk 1]
```diff
@@ -1040,28 +1040,28 @@ EOF
   is_deeply \@li, ['F'], 'found third last li element';
   @li = ();
   $dom->find('li:nth-child(1n+0)')->each(sub { push @li, shift->text });
-  is_deeply \@li, [qw(A B C D E F G)], 'found all li elements';
+  is_deeply \@li, [qw(A B C D E F G H)], 'found all li elements';
```
Comment (kraih) [inline t/mojo/dom.t:1043]: This looks incorrect too. In Chrome i get `A C E F H I` for `li:nth-child(1n+0)`.

Comment (mauke) [reply t/mojo/dom.t:1043]: Can't reproduce, at least with Firefox. I put
```html
<style>
    li:nth-child(1n+0) {
        outline: 1px solid green;
    }
</style>
<ul>
    <li>A</li>
    <li>B</li>
    <li>C</li>
    <li>D</li>
    <li>E</li>
    <li>F</li>
    <li>G</li>
    <li>H</li>
</ul>
```
in a local file and opened it in the browser. All list items got a green border. Also, running `document.querySelectorAll('li:nth-child(1n+0)').length` in the developer console gave `8`.

Comment (mauke) [reply t/mojo/dom.t:1043]: Oh, I see what you did. You accidentally took the HTML code from line 1077, which is part of a different test. The HTML code you used has `<p>B</p>`, `<p>D</p>`, and `<p>G</p>`, which is why they don't show up with a `li` selector, and it has an additional `<li>I</li>` element. The correct HTML code for this test is in line 920 and doesn't contain any `I`.

Comment (kraih) [reply t/mojo/dom.t:1043]: Oops, that's exactly what i did.

**Review by kraih — APPROVED** [review]

**Comment by kraih** [comment]
[@mojojs/dom](https://github.com/mojolicious/dom.js) will have the same bug if you want to apply the fix there too.

**Comment by kraih** [comment]
Since you didn't say anything, i've ported the fix over to `@mojojs/dom`. https://github.com/mojolicious/dom.js/commit/a7ac95d131f2b4c612bdbe5db902a788154c1b80

**Pull request merged**